src/transformers/models/realm/modeling_realm.py

```diff
@@ -1737,11 +1742,59 @@ def __init__(self, config, retriever=None):
         self.post_init()
 
     @property
-    def beam_size(self):
+    def searcher_beam_size(self):
         if self.training:
             return self.config.searcher_beam_size
         return self.config.reader_beam_size
 
+    def block_embedding_to(self, device):
+        """Send `self.block_emb` to a specific device. This is useful for accelerating training and inference speed.
+
+        Args:
+            device (`str` or `torch.device`):
+                The device that `self.block_emb` will be sent to.
+        """
+
+        self.block_emb = self.block_emb.to(device)
+
+    def to(self, *args, **kwargs):
```

Comment:
I think we should allow the `block_emb` to be on CUDA. If we overwrite `to(...)` here this means that one cannot the full forward pass on GPU. Given that the computation is much faster on GPU for `block_emb`, I don't think we should add such a method here. IMO, it'd be enough to have `block_embedding_to(...)` no?

Reply:
Okay. I tested if we firstly call model.to("cuda") and then call `model.block_embedding_to("cpu")`, the gpu memory occupied by `block_emb` could be released, and the training ran normally on single RTX 2080Ti.

Note that one potential scenario is that if one's gpu has only 6 GiB memory like RTX 2060, the model will not be able to run training/inference on GPU following the above workflow because `model.to("cuda")` will consume roughly 8~9 GiB memory if `block_emb` is also sent to GPU, and the cuda out of memory error will occur.

Instead, if we override `model.to(...)`, we can do the following workflows:

with 40 GiB gpu memory like A100:
```
# training / inference (I think this will do full forward pass on GPU? let me know if I'm wrong)
model.to("cuda")
model.block_embedding_to("cuda")
```

with 11 GiB gpu memory like RTX 2080Ti:
```
# training (memory will not enough to store gradients)
model.to("cuda")

# inference
model.to("cuda")
model.block_embedding_to("cuda")
```

with 6 GiB gpu memory like RTX 2060:

```
# inference (not sure if training is possible)
model.to("cuda") # roughly consume 4 GiB
```

We can probably add FAISS support in the future, which is more compatible with PyTorch than ScaNN.

Reply:
Hey @qqaatw, 

I still don't think it's worth trying to optimize for a potential 6GB GPU. Most GPU's people work with have at least 16GB nowadays IMO. Adding a specific `def to(...)` method is very much not in line with other models in the library and might be pretty confusing to the user (note that 99% of users don't read code). So I would prefer to leave as it is here now :-)

Reply:
OK that makes sense. I've remove it :)